docs/ragged_batching.md

```diff
@@ -0,0 +1,131 @@
+implementation. For backends that are known to support ragged batching
+(ONNX Runtime backend, TensorFlow backend, TensorRT backend), they require
+that the model accepts the inputs as 1-dimensional tensors, and the backends
+will concatenate the ragged request inputs into the 1-dimensional tensor.
+
+Because the concatenated input doesn't track the start and end index for each
+request, the backends also require the model to have additional input(s),
+[batch input](#batch-input), that describe various information about the batch
+formed.
+
+## Batch Input
+
+Batch input is often used in combination with ragged input to provide
+information about the each batch element, such as the element count
+of an input for each request in the batch. A batch input is generated by
+Triton instead of being provided in the request, because the information can
+only be finalized after the dynamic batch is formed.
+
+Besides element count,
+there are other batch input kinds that the user can specify, see the
+[protobuf documentation](https://github.com/triton-inference-server/common/blob/main/protobuf/model_config.proto)
+for details.
+
+## Example on Ragged Input and Batch Input
+
+If you have a model that accepts 1 variable length input tensor, INPUT, with
+shape [ -1, -1 ]. The first dimension is the batch dimension, and the second
+dimension is the variable-length content. When the client sends 3 requests of
+shapes [ 1, 3 ], [ 1, 4 ], [ 1, 5 ]. To exploit dynamic batching, the
+straight-forward way to implement this model would expect INPUT shape [ -1, -1 ]
+and assume that all inputs were padded to same length so that all requests
+become shape [ 1, 5 ] and thus Triton can batch and send them to the model.
+In this case, there will be overhead on padding the tensor and on extra model
+computation on the padded content. Below is the input config:
+
+```
+max_batch_size: 16
+input [
+  {
+    name: "INPUT"
+    data_type: TYPE_FP32
+    dims: [ -1 ]
+  }
+]
+```
+
+With triton ragged batching, the model will be implemented to expect INPUT shape
+[ -1 ] and an additional batch input, INDEX, shape [ -1 ] which the model will
+use interpret the batch elements in INPUT. For such model,
+the client requests don't need to be padded and they can be sent as they are
+(with shapes [ 1, 3 ], [ 1, 4 ], [ 1, 5 ]). The backends discussed above will
+batch the input into a tensor of shape [ 12 ] which contains the 3 + 4 + 5
+concatenation of the requests. Triton also creates the batch input tensor of
+shape [ 3 ] with value [ 3, 7, 12 ] which gives the offset into the input tensor
+where each batch element ends. Below is the input config:
+
+```
+max_batch_size: 16
+input [
+  {
+    name: "INPUT"
+    data_type: TYPE_FP32
+    dims: [ -1 ]
+    allow_ragged_batch: true
+  }
+]
+batch_input [
+  {
+    kind: BATCH_ACCUMULATED_ELEMENT_COUNT
+    target_name: "INDEX"
+    data_type: TYPE_FP32
+    source_input: "INPUT"
+  }
+]
+```
```

Comment:
"The above example uses BATCH_ELEMENT_COUNT (protobuf link) type of ragged batching. Other types described in [link] operate similarly."